zpages structured response (#52720)
kubernetes/website: content/en/docs/reference/instrumentation/zpages.md

- [z-pages](#z-pages)
  - [statusz](#statusz)
    - [structured response](#structured-response)
  - [flagz](#flagz)
    - [structured response](#structured-response-1)

### statusz

> Enabled using the `ComponentStatusz` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/),

Reviewer: ```suggestion
Enabled using the `ComponentStatusz` [feature gate](/docs/reference/command-line-tools-reference/feature-gates#ComponentStatusz),
```

Go version: go1.23.2
Binary version: 1.32.0-alpha.0.1484&#43;5eeac4f21a491b-dirty
Emulation version: 1.32.0-alpha.0.1484
Paths: /healthz /livez /metrics /readyz /statusz /version
```

#### structured response

> {{< feature-state for_k8s_version="v1.35" state="alpha" >}}

Reviewer 2: Do we have a feature gate for this change?

> Starting with Kubernetes v1.35, the `/statusz` endpoint supports a structured, versioned response format when requested with the appropriate `Accept` header. Without an `Accept` header, the endpoint returns the plain text response format by default.

Reviewer 2: ```suggestion
Starting with Kubernetes v1.35, the `/statusz` endpoint supports a structured,
versioned response format when requested with the appropriate `Accept` header.
Without an `Accept` header, the endpoint returns the plain text response format by default.
```

To request the structured response, use:
```

> Accept: application/json;v=v1alpha1;g=config.k8s.io;as=Statusz

Reviewer 2: Would `Accept: application/json` work?

Author: no, the `v`, `g`, `as` params are required.

```

Example structured response:
```json
{
  "kind": "Statusz",
  "apiVersion": "config.k8s.io/v1alpha1",
  "metadata": {
    "name": "kube-apiserver"
  },
  "startTime": "2025-10-29T00:30:01Z",
  "uptimeSeconds": 856,
  "goVersion": "go1.23.2",
  "binaryVersion": "1.35.0-alpha.0.1595+b288caa2c26536-dirty",
  "emulationVersion": "1.35",
  "paths": [
    "/healthz",
    "/livez",
    "/metrics",
    "/readyz",
    "/statusz",
    "/version"
  ]
}
```

> The structured response includes:

Author: Added the godoc for schema

	UptimeSeconds int64 `json:"uptimeSeconds"`
	// GoVersion is the version of the Go programming language used to build the binary.
	// The format is not guaranteed to be consistent across different Go builds.
	// +optional
	GoVersion string `json:"goVersion,omitempty"`
	// BinaryVersion is the version of the component's binary.
	// The format is not guaranteed to be semantic versioning and may be an arbitrary string.
	BinaryVersion string `json:"binaryVersion"`
	// EmulationVersion is the Kubernetes API version which this component is emulating.
	// if present, formatted as "<major>.<minor>"
	// +optional
	EmulationVersion string `json:"emulationVersion,omitempty"`
	// MinimumCompatibilityVersion is the minimum Kubernetes API version with which the component is designed to work.
	// if present, formatted as "<major>.<minor>"
	// +optional
	MinimumCompatibilityVersion string `json:"minimumCompatibilityVersion,omitempty"`
	// Paths contains relative URLs to other essential read-only endpoints for debugging and troubleshooting.
	// +optional
	Paths []string `json:"paths,omitempty"`
}
```

### flagz

> Enabled using the `ComponentFlagz` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/), the `/flagz` endpoint shows you the command line arguments that were used to start a component.

Reviewer: ```suggestion
Enabled using the `ComponentFlagz` [feature gate](/docs/reference/command-line-tools-reference/feature-gates#ComponentFlagz), the `/flagz` endpoint shows you the command line arguments that were used to start a component.
```

authorization-webhook-version=v1beta1
default-watch-cache-size=100
```

#### structured response

> {{< feature-state for_k8s_version="v1.35" state="alpha" >}}

Author: replied.

> Starting with Kubernetes v1.35, the `/flagz` endpoint supports a structured, versioned response format when requested with the appropriate `Accept` header. Without an `Accept` header, the endpoint returns the plain text response format by default.

Reviewer 2: ```suggestion
Starting with Kubernetes v1.35, the `/flagz` endpoint supports a structured,
versioned response format when requested with the appropriate `Accept` header.
Without an `Accept` header, the endpoint returns the plain text response format by default.
```

To request the structured response, use:
```
Accept: application/json;v=v1alpha1;g=config.k8s.io;as=Flagz
```

Example structured response:
```json
{
  "kind": "Flagz",
  "apiVersion": "config.k8s.io/v1alpha1",
  "metadata": {
    "name": "kube-apiserver"
  },
  "flags": {
    "advertise-address": "192.168.8.4",
    "allow-privileged": "true",
    "anonymous-auth": "true",
    "authorization-mode": "[Node,RBAC]",
    "enable-priority-and-fairness": "true",
    "profiling": "true",
    "default-watch-cache-size": "100"
  }
}
```

The structured response includes:
- `kind`: Always "Flagz"
- `apiVersion`: The API version (currently "config.k8s.io/v1alpha1")
- `metadata.name`: The name of the component (e.g., "kube-apiserver", "kube-controller-manager", "kube-scheduler")
- `flags`: A map of all command-line flags and their values

{{< note >}}

> The structured responses for both `/statusz` and `/flagz` are alpha features in v1.35 and are subject to change in future releases. They are intended to provide machine-parseable output for debugging and introspection tools. 

Reviewer 2: ```suggestion
The structured responses for both `/statusz` and `/flagz` are alpha features in v1.35
and are subject to change in future releases.
They are intended to provide machine-parseable output for debugging and introspection tools. 
```

> As these features mature, more stable versions of the API (such as `v1beta1` or `v1`) will be introduced in subsequent Kubernetes releases.

Reviewer 2: This line can be dropped safely.

authorization-webhook-version=v1beta1
default-watch-cache-size=100
```

> #### structured response

Reviewer 3: ```suggestion
#### flagz (structured)
```

{{< feature-state for_k8s_version="v1.35" state="alpha" >}}

Starting with Kubernetes v1.35, the `/flagz` endpoint supports a structured,
versioned response format when requested with the appropriate `Accept` header.
Without an `Accept` header, the endpoint returns the plain text response format by default.

To request the structured response, use:
```
Accept: application/json;v=v1alpha1;g=config.k8s.io;as=Flagz
```

{{< note >}}
If you request `application/json` without specifying all required parameters (`g`, `v`, and `as`),
the server will respond with `406 Not Acceptable`.
{{< /note >}}

Example structured response:
```json
{
  "kind": "Flagz",
  "apiVersion": "config.k8s.io/v1alpha1",
  "metadata": {
    "name": "kube-apiserver"
  },
  "flags": {
    "advertise-address": "192.168.8.4",
    "allow-privileged": "true",
    "anonymous-auth": "true",
    "authorization-mode": "[Node,RBAC]",
    "enable-priority-and-fairness": "true",
    "profiling": "true",
    "default-watch-cache-size": "100"
  }
}
```

> The schema for the structured `/flagz` response is defined in [KEP-4828](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md):

Reviewer: ```suggestion
The schema for the structured `/flagz` response is defined in [KEP-4828](https://git.k8s.io/enhancements/keps/sig-instrumentation/4828-component-flagz/README.md):
```
